pkg/lightning/backend/remote/chunk_sender.go

```diff
@@ -0,0 +1,398 @@
+func (c *chunkSender) putEmptyChunk(ctx context.Context) error {
+	chunk := &chunk{id: c.getLastChunkID(), data: nil}
+	return c.putChunkToRemote(ctx, chunk)
+}
+
+func (c *chunkSender) putChunkToRemote(ctx context.Context, chunk *chunk) error {
+	if len(chunk.data) != 0 {
+		// Cache the chunk data to avoid the data being lost when the remote worker restarts.
+		// If the remote worker restarts, we can retry sending chunks from cache.
+		err := c.chunksCache.put(chunk.id, chunk.data)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	url := fmt.Sprintf(putChunkURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id, chunk.id)
+
+	data, err := sendRequest(ctx, c.httpClient, "PUT", url, chunk.data)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	result := new(PutChunkResult)
+	err = json.Unmarshal(data, result)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = c.handlePutChunkResult(ctx, result, chunk.id)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// handlePutChunkResult handles the result of put chunk request.
+//
+// Chunks are sent to remote workers in sequence according to chunk id. The remote
+// worker should return the `FlushedChunkID` and `HandledChunkID`. If the remote worker
+// restarts, the `FlushedChunkID` may be less than the `expectedChunkID`, and remote
+// worker will reject the new chunk. In this case, we need to retry sending chunks
+// from the `FlushedChunkID` to the `expectedChunkID`.
+func (c *chunkSender) handlePutChunkResult(ctx context.Context, result *PutChunkResult, expectedChunkID uint64) error {
+	if result.Canceled {
+		c.logger.Error("failed to put chunk, task is canceled", zap.String("error", result.Error))
+		return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+	}
+
+	if result.HandledChunkID != expectedChunkID {
+		c.logger.Info("remote worker may restart, retry to put chunk",
+			zap.Uint64("expected chunkID", expectedChunkID),
+			zap.Uint64("handled chunkID", result.HandledChunkID),
+			zap.Uint64("flushed chunkID", result.FlushedChunkID))
+
+		nextChunkID := result.HandledChunkID + 1
+		for nextChunkID <= expectedChunkID {
+			url := fmt.Sprintf(putChunkURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id, nextChunkID)
+
+			buf, err := c.chunksCache.get(nextChunkID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			c.logger.Info("retry to put chunk", zap.Uint64("chunkID", nextChunkID))
+
+			data, err := sendRequest(ctx, c.httpClient, "PUT", url, buf)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			result = new(PutChunkResult)
+			err = json.Unmarshal(data, result)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if result.Canceled {
+				c.logger.Error("failed to put chunk, task is canceled", zap.String("error", result.Error))
+				return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+			}
+
+			nextChunkID = result.HandledChunkID + 1
```

Comment:
if worker restarts again, we should go back to line 283.

Reply:
We don't need back to line 283, because the `nextChunkID` changes based on the `HandledChunkID`, thus ensuring that we can resend the chunk from `HandledChunkID` to `expectedChunkID`.